### PR TITLE
Quartz - make it possible to access the underlying scheduler instance

### LINKED ISSUE #7246: Quartz extension - make it possible to access the underlying Quartz scheduler instance
We should probably just add a producer for `org.quartz.Scheduler` and make it injectable.

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #7246

### CODE REVIEW

File: docs/src/main/asciidoc/quartz.adoc [hunk 1]
```diff
@@ -124,6 +124,58 @@ public class TaskBean {
...
+    void performTask() {
+        Task task = new Task();
+        task.persist();
+    }
+    
+    // A new instance of MyJob is created by Quartz for every job execution
+    static class MyJob implements Job {
+    
+       public void execute(JobExecutionContext context) throws JobExecutionException {
+          Arc.container().instance(TaskBean.class).get().perform(); <3>
```
Comment (machi1990) [inline docs/src/main/asciidoc/quartz.adoc]:
```suggestion
          Arc.container().instance(TaskBean.class).get(). performTask(); <3>
```

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java [hunk 2]
```diff
@@ -19,4 +19,12 @@
     @ConfigItem(defaultValue = "5")
     public int threadPriority;
 
+    /**
+     * By default, the scheduler is not started unless a {@link io.quarkus.scheduler.Scheduled} business method is found.
+     * If set to true the scheduler will be started even if no scheduled business methods are found. This is necessary for
+     * "pure" programmatic scheduling.
+     */
+    @ConfigItem(defaultValue = "false")
```
Comment (machi1990) [inline extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java]: `defaultValue` is not necessary. `boolean` config items defaults to `false`

File: docs/src/main/asciidoc/quartz.adoc [hunk 3]
```diff
@@ -124,6 +124,58 @@ public class TaskBean {
...
+
+It is also possible to leverage the Quartz API directly.
+You can inject the underlying `org.quartz.Scheduler` in any bean:
+
+[source,java]
+----
+package org.acme.quartz;
+
+@ApplicationScoped
+public class TaskBean {
```
Comment (machi1990) [inline docs/src/main/asciidoc/quartz.adoc]: We should probably update [Quartz QS](https://github.com/quarkusio/quarkus-quickstarts/tree/development/quartz-quickstart)  with this example

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/InjectQuartzSchedulerTest.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+import io.quarkus.runtime.StartupEvent;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class InjectQuartzSchedulerTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(Starter.class)
+                    .addAsResource(new StringAsset("quarkus.quartz-runtime.force-start=true"),
```
Comment (machi1990) [inline extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/InjectQuartzSchedulerTest.java]:
```suggestion
                    .addAsResource(new StringAsset("quarkus.quartz.force-start=true"),
```

**Review by machi1990 — CHANGES_REQUESTED** [review]
@mkouba I have added a couple of suggestions. We should also update the QS code.

File: docs/src/main/asciidoc/quartz.adoc (hunk 3 above)
Comment (mkouba) [reply docs/src/main/asciidoc/quartz.adoc]: I don't think so. This is more like a corner case, ie. something we don't want to push...

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java (hunk 2 above)
Comment (mkouba) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzRuntimeConfig.java]: I prefer explicit values so I'd like to keep it as it is.

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/InjectQuartzSchedulerTest.java (hunk 4 above)
Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/InjectQuartzSchedulerTest.java]: Unfortunately, `quarkus.quartz.force-start=true` would not work. See https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/.60RuntimeConfig.60.20vs.20.60RunTimeConfig.60

Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/InjectQuartzSchedulerTest.java]: I'm going to send a PR to fix https://github.com/quarkusio/quarkus/issues/7257 - afterwards, it should be possible to switch to `quarkus.quartz.force-start=true`.

Comment (machi1990) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/InjectQuartzSchedulerTest.java]: Thanks @mkouba

**Review by machi1990 — APPROVED** [review]
LGTM. Thanks @mkouba

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java [hunk 5]
```diff
@@ -56,8 +59,15 @@
     private final AtomicInteger triggerNameSequence;
     private final Map<String, ScheduledInvoker> invokers;
 
+    @Produces
+    @Singleton
```
Comment (gastaldi) [inline extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java]: Should we use `@DefaultBean` here?

Comment (mkouba) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java]: I don't think so. Why?

Comment (gastaldi) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java]: When the user code `@Produces` their own `org.quartz,Scheduler`/?

Comment (gastaldi) [reply extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java]: But maybe we shouldn't allow that anyway :)

**Comment by mkouba** [comment]
@gsmet do know how to restart the checks? The CI seems to be stuck.

**Comment by gastaldi** [comment]
@mkouba try forcing push, worked for me

**Comment by machi1990** [comment]
Seems like there is a timeout failure `The job running on runner GitHub Actions 40 has exceeded the maximum execution time of 35. `

**Comment by gsmet** [comment]
Yeah we need to adjust things, just building the `kubernetes-client` native image takes 15mn.

**Comment by gsmet** [comment]
I added a commit here to reorganize the categories: https://github.com/quarkusio/quarkus/pull/7312 .

**Comment by gsmet** [comment]
The Quartz module was successfully built so let's merge.

**Comment by machi1990** [comment]
Thanks @gsmet

**Pull request merged**